Add New Plugin P101, Wake On LAN [Testing]

thomastech opened the pull request:
**New Plugin P101, Wake On LAN [Testing]**
Summary: Wake On LAN (WOL) Sends a "Magic Packet" to a computer's MAC address. It Wakes-up a sleeping computer if it has been configured for WOL.

This PR includes the readthedocs files for the new P101 plugin.

For background story please see https://www.letscontrolit.com/forum/viewtopic.php?f=4&t=8010

Update: This PR also includes minor edits to the P075 Nextion and P079 Wemos/LoLin MotorShield Docs (readthedocs).

- Thomas

thomastech committed:
Merge pull request #4 from letscontrolit/mega

Sync Local Repo

thomastech committed:
Added Wake On LAN Plugin P101

Wake On LAN (WOL) Plugin. 
Sends Magic Packet to MAC address; Wake-up sleeping computer if configured for WOL.

thomastech committed:
Updated Command.rst for Wake On LAN Plugin P101

thomastech committed:
Added Doc files for Wake On LAN Plugin P101

Updated ReadtheDocs for the new Wake On LAN (WOL) Plugin P101.

thomastech committed:
Update _plugin_categories.repl

Removed extra comma at line containing ref:`P101_page`

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+      //  Warning, event->TaskIndex is incorrectly set in PLUGIN_WRITE if plugin is disabled.
+      //  So checking the task enable status needs special attention.
+      //  See https://github.com/letscontrolit/ESPEasy/issues/3317
+      if (event->TaskIndex < TASKS_MAX) {
```
See `validTaskIndex` (src/Globals/Plugins.cpp)

thomastech replied:
Ahhh. Exactly the same thing, but in a pretty function. I'll update the code using it.

thomastech replied:
Updated file, see line 298.

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+  char   ipString[IP_BUFF_SIZE_P101]   = "";
+  char   macString[MAC_BUFF_SIZE_P101] = "";
+  String msgStr;
+  const String wolStr = F(LOG_NAME_P101);
```
Please try to allocate those only when needed, otherwise they are also created each time the plugin is called like in `PLUGIN_FIFTY_PER_SECOND` even though this plugin doesn't even handle such a case.

thomastech replied:
I'll correct this.
- Thomas

thomastech replied:
Update file. Allocations now done in local scope for PLUGIN_WEBFORM_LOAD, PLUGIN_WEBFORM_SAVE, and PLUGIN_WRITE.
- Thomas

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+        LoadCustomTaskSettings(event->TaskIndex, strings, 2, CUSTOMTASK_STR_SIZE_P101);
+        safe_strncpy(ipString,  strings[0],  IP_BUFF_SIZE_P101);
+        safe_strncpy(macString, strings[1], MAC_BUFF_SIZE_P101);
+      }
```
You try to load the settings even when the command is not valid.
So on every command processed by ESPEasy (that was not yet processed by a taskIndex lower than this one) the extra tasksettings are loaded.
If you need to read this from a number of places, then you can make a function for it or do a check first to see if the command is one that can be executed by this task.

thomastech replied:
I'll correct this.

thomastech replied:
Updated File, LoadTaskSettings() now done after qualifying command keyword. See line 301.

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+      //  So checking the Device Name needs special attention.
+      //  See https://github.com/letscontrolit/ESPEasy/issues/3317
+      if (cmd.equalsIgnoreCase(F(CMD_NAME_P101)) ||
+          ((event->TaskIndex < TASKS_MAX) && cmd.equalsIgnoreCase(getTaskDeviceName(event->TaskIndex)))) {
```
`validTaskIndex`

And this is also a bit tricky: `cmd.equalsIgnoreCase(getTaskDeviceName(event->TaskIndex)`
So you're effectively allowing to call a command by a task name.
So if I name a task "reboot", it is a bit hard to predict what will happen.

thomastech replied:
The plugin purposely allows using the task name as the command. But you are correct, someone could very likely name their WOL task as REBOOT.  I will add a test in PLUGIN_WEBFORM_SAVE that prevents using REBOOT as the task name.

TD-er replied:
Maybe also 'reset' :)

I guess we should add (later) a quick check to see if a word is a reserved keyword, based on the included plugins.
This could also speed-up command processing quite a bit.

Related PR: https://github.com/letscontrolit/ESPEasy/pull/3198

thomastech replied:
Got it. I'll include RESET in the reserved naming test.

thomastech replied:
Updated File, see safeName() function, Line 419.
- Thomas

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+        success = true;
+
+        // Do not process WOL command if the plugin is disabled.
+        if ((taskEnable == false) || (event->TaskIndex >= TASKS_MAX)) {
```
`if (!taskEnable || !validTaskIndex(event->TaskIndex)) {`

... which is the same as:
`if (!(taskEnable && validTaskIndex(event->TaskIndex))) {`

thomastech replied:
I didn't mean to disrespect Mr. Karnaugh. I will change it.
- Thomas

TD-er replied:
Just a fun fact...
When I had the boolean algebra course at university, I had the feeling it was less useful than watching grass grow and paint dry at the same time... like who would ever use that !?!

Now when I see complex logic statements, it is like they start glowing to me and I need to have a look.

thomastech replied:
Updated file. After a bit of thought I was able to further simplify the !taskEnable related code. See line 305.
- Thomas

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+
+  if ((length < IP_MIN_SIZE_P101) || (length > IP_ADDR_SIZE_P101)) {
+    return false;
+  }
```
Just curious, is it needed to check for the string length?
Is it possible the `ip.fromString` may return true on strings that do contain more or less characters?
What about IPv6?

thomastech replied:
1. ip.fromString does not fully test the IP length, so some irregular cases can get through. Learned that during validation. So I check the length for irregular length before calling ip.fromString().
2. IPv6 is not supported by the WOL plugin.

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+// validateMac(): MAC Address validity checker.
+// Arg: MAC Address String in colon separated format (00:00:00:00:00:00).
+// Return true if MAC string appears legit.
+bool validateMac(String macStr) {
```
On some OS'es the leading 0 in a hex field may be left out.  (e.g. Mac OS)
Copy-paste of those MAC addresses will thus be rejected.

thomastech replied:
During submission, the user will see a message in the plugin that reports any problems with the MAC formatting. This keeps things simple. 

So if a copied MAC string is not formatted as aa:bb:cc:dd:ee:ff then the user will be warned. And will need to manually add the missing or illegal chars and re-submit.

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -0,0 +1,475 @@
+
+  portNumber = portStr.toInt();
+
+  if ((portNumber < 0) || (portNumber > PORT_MAX_P101)) {
```
Is portnr 0 a valid port?
That's the return value of `toInt()` when it is a string with non-numerical characters.

thomastech replied:
Port 0 is valid. Regarding non-numerics, there's a very simple isDigit() test at line 466.  This only checks the leading char, which is enough to find blatant port value mistakes in the command sentence.

thomastech committed:
Updated ReadTheDocs for P075 Nextion.

Minor  docs update for P075 Nextion.

thomastech committed:
Update P075 and P079 Plugin Docs

Minor ReadTheDocs revisions to P075 Nextion Display and P079 Wemos/Lolin Motor Shield.

TD-er commented on docs/source/Plugin/P075.rst:
```diff
@@ -276,7 +277,7 @@ It also sends new touch event values to the controller.
 
-Nextion Command Format Summary
+Nextion Data Format Summary
 ------------------------------
```
Make sure the line (to signal the (sub)section level) has the same number of characters as the title itself.

thomastech replied:
I'll need a bit more hand holding on this since your comment is unclear.

Do you mean to change this:
```
Nextion Data Format Summary
------------------------------
```
To this:
```
Nextion Data Format Summary
---------------------------
```

If that is the issue then the existing text is suppose to be OK. The heading rules seem to confirm the extra line length is OK:
```
If under and overline are used, their length must be identical
The length of the underline must be at least as long as the title itself
```
More details here:
https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html

TD-er replied:
OK, didn't know it could also be longer.
I noticed in the docs that Sphinx gave warnings when the underline was shorter.

Learned something myself today :)

thomastech replied:
> Learned something myself today 

I'm new to reStructured Text, so avoid using me as a trusted teacher. :)

I like rst's pretty output. But getting it is like using the word processor on an old PDP-11 mainframe while typing on a VT100 terminal.

TD-er replied:
Well programming the ESP isn't that different.
Amount of memory is in the same order of magnitude and debugging is still very basic.
Only main improvement is that when you let your code slip out of your hand you don't have to get on your knees to collect all punch cards again and re-order them.

I've started programming a MSX-1 from 1983, when I was 7 and back then I had more free memory compared to what we now have on the ESP8266.
But instead of mounting Lego, paperclips and aluminum foil covered pins connected to the joystick ports to count pulses, we now use slightly more fancy switches to count and the connectors are less clunky :)
So there has been some progress in 37 years I've been working with computers, electronics and programming. 👍

thomastech committed:
Update _P101_WakeOnLan.ino

Made changes as requested. Will update the review comments with more info.

TD-er commented on src/_P101_WakeOnLan.ino:
```diff
@@ -95,21 +98,17 @@ WakeOnLan WOL(udp);
+uint8_t safeName(taskIndex_t index);
+bool    validateIp(String ipStr);
+bool    validateMac(String macStr);
+bool    validatePort(String portStr);
```
Please use `const String&` as function argument.
By giving the argument as a const reference, you prevent the content to be copied.
As it is now, you copy the content into a new object.

thomastech replied:
Nice! New file uploaded.
- Thomas

TD-er commented:
The WOL library is not included in this PR (in the lib folder)

thomastech committed:
Add WakeOnLan V1.1.6 Library

thomastech committed:
Update _P101_WakeOnLan.ino

Per TD-er request, changed calling argument from String to const String& on three functions.

thomastech commented:
This PR includes some minor edits to the P075 Nextion and P079 Wemos/Lolin Motor shield ReadTheDocs rst files (3 files). I intended to create a separate PR for this. But now I know if I commit something to my local working branch and a PR is open, the changes automatically get added to the open PR over here. Cool automation, just wasn't expecting it.

- Thomas

TD-er committed:
Merge branch 'mega' into feature/3328_WOL

TD-er commented:
@thomastech About the PR and Git branches stuff...
It is best to create a branch to start hacking and then commit to that branch.
When ready to create a pull request, you make a request to merge your branch into the main branch (we call it "mega" here)
So when you push new commits to the same branch, they are automatically added to the same PR, as it is still the same branch you want to merge.

For this PR, you didn't create a new branch, but are still working on your version of the "mega" branch.
So I will now merge this or else you will get into trouble when you try to get your stuff in sync due to my just merged "really big PR".
So as tip for future PRs, please create a local branch on your clone of the repository and work on there.

thomastech commented:
@TD-er, thank you for merging P101. I appreciate your code review and all the suggested improvements.

- Thomas

TD-er commented:
You're welcome :)
